### PR TITLE
fix: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions

### LINKED ISSUE #1689: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions
## Problem

Entity enrichment extracts mentions from fact text and writes them to `fact_entity_mentions`. Current output shows quality problems:

### Noise examples observed (2026-05-29 job 3)

1. **Platform/service names misclassified as ORG**:
   - `ORG:"whatsapp"` — should be a platform tag, not an org
   - `ORG:"hybrid-memory plugin"` — free text, not a named entity

2. **Truncated/partial extractions**:
   - `ORG:"sessio"` — clearly truncated from "session" or similar
   - `PERSON:"Surgeon"` repeated 3× — likely a role title, not a person name

3. **Duplicate same-fact repeats**:
   - `Hybrid-memory PR Pipeline` repeated 4× in one fact's enrichment output
   - Same entity listed multiple times per fact

4. **Case inconsistency**:
   - `Hybrid-memory` vs `hybrid-memory` treated as different entities

## Expected behaviour

Before writing to `fact_entity_mentions`:

1. **Normalize**: lowercase comparison, trim whitespace, collapse runs of whitespace
2. **Deduplicate**: one entry per `(fact_id, label, normalized_text)` — use `INSERT OR IGNORE` or equivalent
3. **Filter noise**: 
   - Drop single-character or very short mentions (< 3 chars)
   - Drop common English/system words not in an actual entity list
   - Drop mentions that are substrings of longer valid mentions in the same fact
4. **Classify type better**:
   - `whatsapp`, `signal`, `telegram` → platform/service tag, not ORG
   - `Hybrid-memory plugin`, `the agent`, `the system` → no entity
   - Roles/titles like `Surgeon`, `Engineer`, `Manager` without a surname → no PERSON

## Evidence

Job 3 enrich-entities output showed:
```
26ca2b8e... ORG:"Google", ORG:"Gemini", ORG:"MiniMax"   ← good
2bb1765c... ORG:"MiniMax"                                  ← good
[suspicious entries redacted above]
```

DB query post-job 3:
```sql
SELECT label, COUNT(*) FROM fact_entity_mentions GROUP BY label;
-- ORG: 55, PERSON: 32
-- Some entries clearly noise quality
```

## Labels
labels: bug, priority:medium

### PR DESCRIPTION
**Pull request by markus-lassfolk**
Issue: https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1689

Status: implementation completed on the current PR branch.
Protection: keep `stage/implementing` until Issue Stewardship dispatches/receives implementation and explicitly hands off to PR Stewardship.

Enrichment present on issue: 1
Priority inherited from issue: Medium (source labels: priority:medium)
Dependencies/blocked labels inherited from issue: none

Implementation PR created by Issue Stewardship. Copilot implementation now adds the issue-scoped persistence change and keeps behavior outside the acceptance criteria unchanged.

---

&gt; [!NOTE]
&gt; **Medium Risk**
&gt; Touches entity/mention persistence in the memory pipeline; normalization, deduplication, and filtering can affect downstream fact linkage if wrong.
&gt; 
&gt; **Overview**
&gt; This PR updates **entity enrichment** so that mentions are normalized before persistence, deduplicated per fact/label/normalized mention, and filtered for short or generic noise before being written to **`fact_entity_mentions`**.
&gt; 
&gt; The implementation is focused on the persistence path and includes focused tests plus direct docs updates for the mention persistence behavior.
&gt; 
&gt;<sup>Reviewed by <a href="https://cursor.com/bugbot">Cursor Bugbot</a> for commit a23893eca32fa43ba442661e65a7d83e8abe132c. Bugbot is set up for automated code reviews on this repo. Configure <a href="https://www.cursor.com/dashboard/bugbot">here</a>.</sup>

---

> [!NOTE]
> **Medium Risk**
> Changes what contacts, organizations, and org_fact_links get created from enrichment; incorrect filtering or containment logic could drop real entities or merge distinct ones, though behavior is heavily unit-tested.
> 
> **Overview**
> Entity enrichment now runs **PERSON**/**ORG** mentions through `normalizeFactEntityMentionsForPersistence` in `replaceFactEntityMentions` before SQLite writes, instead of persisting raw LLM spans.
> 
> The pipeline **trims** surfaces and **fixes offsets**, **drops** very short or generic noise (stop words, chat apps like WhatsApp, lone role titles), **removes** shorter spans contained in overlapping longer ones (with punctuation boundaries), and **dedupes** per fact/label/normalized key while keeping the best confidence/length. **GRAPH-MEMORY.md** and the extension README describe this behavior; **entity-layer** tests cover edge cases and an end-to-end persistence check.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9ee12f421b3b13c352527cf95dc695e92ea7ba64. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

Closes #1689

### CODE REVIEW

**Comment by markus-lassfolk** [comment]
@copilot implement this issue on the current PR branch.

Repo/Issue: markus-lassfolk/openclaw-hybrid-memory#1689 https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1689
PR: #1740 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740
Branch: issue-1689-entity-enrichment-should-normalize-and-deduplica
Head: a23893eca32fa43ba442661e65a7d83e8abe132c
Draft: false
Labels: size/XS,stage/implementing,priority:medium

Implementation brief:
Use the issue and PR body as the implementation brief. If acceptance is unclear, report the exact blocker instead of broadening scope.

Constraints:
- Push to this existing PR branch; do not open a separate PR.
- Keep scope focused on issue #1689: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions
- Preserve current behavior outside the acceptance criteria.
- Add/update tests and docs as described in the enrichment.
- If acceptance is unclear or blocked, comment with the exact blocker.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
Review-thread repair dispatch marker for current head (audit breadcrumb only).

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1740 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740
Title: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions (#1689)
Head SHA: 30c5ca690fcbcb53678df95219db2a37f4a83e3f
Branch: issue-1689-entity-enrichment-should-normalize-and-deduplica
Unresolved review threads: 3
Merge state at dispatch: BEHIND/MERGEABLE
Check blockers at dispatch: 1 (checks_failed)
Bridge artifact: /tmp/review-thread-repair-brief-markus-lassfolk_openclaw-hybrid-memory-pr1740-30c5ca690fcb.json

next_owner=Stewardship cron/main (sessions_spawn bridge); wait_until=repair worker evidence + review-thread/merge-check read-back.

**Comment by markus-lassfolk** [comment]
**PR #1740 thread repair — 3 threads resolved**

**Commit:** `0592dc1b` pushed to `issue-1689-entity-enrichment-should-normalize-and-deduplica`

**Fixes applied to** `entity-layer.ts`:

1. **Line 67 — Short mention length filter (`&&` → `||`)**
   `shouldFilterEntityMention` now filters when *either* `normalizedSurface` or `surfaceText` is below 3 chars (was requiring both to be short).

2. **Line ~150 — Offsets not updated after trim**
   Added `endOffset: mention.endOffset - trailingTrim` adjustment after `normalizeMentionSurfaceText`, so stored offsets always match the trimmed `surfaceText` in the original fact text.

3. **Line ~145 — Dedup key uses surface canonical**
   Changed dedup key from `normalizeEntityKey(mention.normalizedSurface)` to `normalizeEntityKey(mention.surfaceText)` — now matches what `upsertOrganization`/`upsertContact` compute for `canonical_key`.

**Tests:** `vitest run tests/entity-layer.test.ts` → 10/10 passed ✅

```
✓ tests/entity-layer.test.ts (10 tests) 250ms
```

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 1]
```diff
@@ -45,6 +46,126 @@ export function normalizeEntityKey(name: string): string {
...
+function shouldFilterEntityMention(mention: {
+  label: EntityMentionLabel;
+  surfaceText: string;
+  normalizedSurface: string;
+}): boolean {
+  if (
+    mention.normalizedSurface.length < MIN_ENTITY_MENTION_LENGTH &&
+    mention.surfaceText.length < MIN_ENTITY_MENTION_LENGTH
+  ) {
+    return true;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts:107]: **Resolved:** `&&` changed to `||` in `shouldFilterEntityMention` — filter now applies when *either* `normalizedSurface` or `surfaceText` is below 3 chars. Commit `0592dc1b`. Tests: 10/10 entity-layer passed.

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 2]
```diff
@@ -45,6 +46,126 @@ export function normalizeEntityKey(name: string): string {
...
+  const prepared = mentions
+    .map((mention) => {
+      const surfaceText = normalizeMentionSurfaceText(mention.surfaceText);
+      const normalizedSurface =
+        normalizeEntityKey(mention.normalizedSurface || surfaceText) || normalizeEntityKey(surfaceText);
+      return {
+        ...mention,
+        surfaceText,
+        normalizedSurface,
+      };
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts:204]: **Resolved:** `endOffset` now adjusted by `trailingTrim` after `normalizeMentionSurfaceText` so stored offsets always match the trimmed `surfaceText` in the original fact text. Commit `0592dc1b`. Tests: 10/10 entity-layer passed.

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 3]
```diff
@@ -45,6 +46,126 @@ export function normalizeEntityKey(name: string): string {
...
+    startOffset: number;
+    endOffset: number;
+    confidence: number;
+  },
+>(mentions: readonly T[]): T[] {
+  const prepared = mentions
+    .map((mention) => {
+      const surfaceText = normalizeMentionSurfaceText(mention.surfaceText);
+      const normalizedSurface =
+        normalizeEntityKey(mention.normalizedSurface || surfaceText) || normalizeEntityKey(surfaceText);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts]: **Resolved:** Dedup key changed to `normalizeEntityKey(mention.surfaceText)` to match what `upsertOrganization`/`upsertContact` compute for `canonical_key`. Commit `0592dc1b`. Tests: 10/10 entity-layer passed.

**Comment by markus-lassfolk** [comment]
@copilot please repair this bounded blocker on the current PR head and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1740 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740
Title: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions (#1689)
Head SHA: df7122ecc5bc57c415c7d7d81f26c2fccdd6adfd
Branch: issue-1689-entity-enrichment-should-normalize-and-deduplica
Current merge state: MERGEABLE/CLEAN
Labels: documentation,extensions,size/L,stage/implementing,priority:medium
Blocker category: unknown
Blocker signature: 730062c4cd3a (dirty-conflict-or-ci)
Attempt: 1/2 for this blocker lineage

Blockers:
- MERGEABLE/CLEAN (unknown)

Requested work:
- Fix the listed blocker(s) on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, verification run, and any remaining blockers.

next_owner=Copilot; escalation_tier=copilot1; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
COUNCIL_VERDICT head=46fa31a780d8b00c1bd94eabde48ef0410f41557 verdict=APPROVE promote=stage/review risk=medium

Quality assessment:
- The core fix is sound: `replaceFactEntityMentions` now normalizes mention rows before persistence, filters short/generic/service noise, deduplicates by canonical surface per label, prefers higher-confidence/longer/earlier mentions, and preserves stable ordering by offset.
- The follow-up fixes addressed the important edge cases I would have worried about: character-substring false positives (`Ford` vs `Oxford`), all-occurrence containment checks, normalized-value abbreviations breaking containment (`Acme` vs `Acme Corporation`), and matching the dedup key to the `upsertOrganization`/`upsertContact` canonical key.
- Tests are substantial and cover pure normalization plus SQLite persistence, org/contact side effects, service-term filtering, role-title filtering, substring preservation, and deduplication behaviour. Documentation was updated in both graph docs and the extension README.

Verification performed:
- `gh pr view 1740 --repo markus-lassfolk/openclaw-hybrid-memory --json body,title,files,additions,deletions,headRefOid` returned head `46fa31a780d8b00c1bd94eabde48ef0410f41557` with changes in `docs/GRAPH-MEMORY.md`, `extensions/memory-hybrid/README.md`, `extensions/memory-hybrid/backends/facts-db/entity-layer.ts`, and `extensions/memory-hybrid/tests/entity-layer.test.ts`.
- Reviewed the full PR patch for the entity normalization/deduplication logic.
- Checked out the PR at the stated head and ran `npm ci` in `extensions/memory-hybrid`.
- `npm test -- tests/entity-layer.test.ts` passed: 1 file, 10 tests.
- `npm run build` completed successfully.

Risk level: medium.
- This touches the entity graph write path and intentionally changes what gets persisted, so behavioural risk is real.
- The implementation is localized, backwards-compatible at the schema/API level, and covered by targeted tests, which keeps the merge risk acceptable.
- One minor non-blocking caveat: offsets are corrected for trailing whitespace after surface trimming, but leading/internal whitespace normalization can still make stored offsets imperfect for deliberately malformed input. The persistence tests do not currently assert offsets. Given NER spans should normally be exact source spans, I do not consider this a blocker for this PR, but a follow-up offset-focused test would be prudent.

Recommendation:
- Promote from `stage/council` to `stage/review` / next pipeline review step.
- No blocking changes requested.

**Comment by markus-lassfolk** [comment]
@codex

**Comment by markus-lassfolk** [comment]
Review-thread repair dispatch marker for current head (audit breadcrumb only).

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1740 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740
Title: fix: entity enrichment should normalize and deduplicate mentions before writing fact_entity_mentions
Head SHA: 419d8d7ba8d5ad7100f77125f8d997e3a6562c23
Branch: issue-1689-entity-enrichment-should-normalize-and-deduplica
Unresolved review threads: 3
Merge state at dispatch: BEHIND/MERGEABLE
Check blockers at dispatch: 0 (none)
Bridge artifact: /tmp/review-thread-repair-brief-markus-lassfolk_openclaw-hybrid-memory-pr1740-419d8d7ba8d5.json

next_owner=Stewardship cron/main (sessions_spawn bridge); wait_until=repair worker evidence + review-thread/merge-check read-back.

**Comment by markus-lassfolk** [comment]
Review-thread repair check on current PR head `419d8d7ba8d5ad7100f77125f8d997e3a6562c23`:

- Verified unresolved thread blockers and code/test state in `extensions/memory-hybrid/backends/facts-db/entity-layer.ts` + `extensions/memory-hybrid/tests/entity-layer.test.ts`.
- The three blocker themes are already implemented on this head:
  - preserve extractor-provided normalized surfaces
  - allow valid two-character entity mentions
  - treat no-space scripts as containment boundaries
- Validation command run:
  - `cd extensions/memory-hybrid && npm test -- tests/entity-layer.test.ts`
  - result: PASS (17/17 tests)

No code changes were required, so no commit/push was made.

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 4]
```diff
@@ -63,6 +64,141 @@ export function normalizeEntityKey(name: string): string {
...
+    normalizedSurface: string;
+    startOffset: number;
+    endOffset: number;
+    confidence: number;
+  },
+>(mentions: readonly T[]): T[] {
+  const prepared = mentions
+    .map((mention) => {
+      const surfaceText = normalizeMentionSurfaceText(mention.surfaceText);
+      const normalizedSurface = normalizeEntityKey(surfaceText);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts]: Verified on live head 419d8d7ba8d5ad7100f77125f8d997e3a6562c23 in fresh /tmp checkout. No new code delta was needed: the current head already contains the requested guard/test coverage in `extensions/memory-hybrid/backends/facts-db/entity-layer.ts` and `extensions/memory-hybrid/tests/entity-layer.test.ts`. Relevant check rerun: `cd extensions/memory-hybrid && npm test -- tests/entity-layer.test.ts` -> PASS, 17/17 tests. Evidence comment: https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740#issuecomment-4586776296

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 5]
```diff
@@ -63,6 +64,141 @@ export function normalizeEntityKey(name: string): string {
...
+}
+
+function shouldFilterEntityMention(mention: {
+  label: EntityMentionLabel;
+  surfaceText: string;
+  normalizedSurface: string;
+}): boolean {
+  if (
+    mention.normalizedSurface.length < MIN_ENTITY_MENTION_LENGTH ||
+    mention.surfaceText.length < MIN_ENTITY_MENTION_LENGTH
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts]: Verified on live head 419d8d7ba8d5ad7100f77125f8d997e3a6562c23 in fresh /tmp checkout. No new code delta was needed: the current head already contains the requested guard/test coverage in `extensions/memory-hybrid/backends/facts-db/entity-layer.ts` and `extensions/memory-hybrid/tests/entity-layer.test.ts`. Relevant check rerun: `cd extensions/memory-hybrid && npm test -- tests/entity-layer.test.ts` -> PASS, 17/17 tests. Evidence comment: https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740#issuecomment-4586776296

File: extensions/memory-hybrid/backends/facts-db/entity-layer.ts [hunk 6]
```diff
@@ -63,6 +64,163 @@ export function normalizeEntityKey(name: string): string {
...
+  }
+
+  // Check if spans overlap: they must overlap for containment to apply
+  const spansOverlap = mention.startOffset < other.endOffset && other.startOffset < mention.endOffset;
+  if (!spansOverlap) {
+    return false;
+  }
+
+  const isWordBoundary = (char: string): boolean => {
+    return char === " " || /[,\-.\/()\[\]{}:;!?'"&]/.test(char);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/entity-layer.ts:141]: Verified on live head 419d8d7ba8d5ad7100f77125f8d997e3a6562c23 in fresh /tmp checkout. No new code delta was needed: the current head already contains the requested guard/test coverage in `extensions/memory-hybrid/backends/facts-db/entity-layer.ts` and `extensions/memory-hybrid/tests/entity-layer.test.ts`. Relevant check rerun: `cd extensions/memory-hybrid && npm test -- tests/entity-layer.test.ts` -> PASS, 17/17 tests. Evidence comment: https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1740#issuecomment-4586776296
